### PR TITLE
Aws sdk can be slow when get ec2 metadata in Aws::Client::ClientConfiguration

### PR DESCRIPTION
**Pull request by caneGuy**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

When we init ClientConfiguration with default constructor, it will try to get current region use ec2 metadata.And there is an exclusive lock in the sdk code which will cause query slow when we have many files to scan.
```
ClientConfiguration::ClientConfiguration()
{
    setLegacyClientConfigurationParameters(*this);
    retryStrategy = InitRetryStrategy();

    if (region.empty() &&
        Aws::Utils::StringUtils::ToLower(Aws::Environment::GetEnv("AWS_EC2_METADATA_DISABLED").c_str()) != "true")
    {
        auto client = Aws::Internal::GetEC2MetadataClient();
        if (client)
        {
            region = client->GetCurrentRegion();
        }
    }
    if (!region.empty())
    {
        return;
    }
    region = Aws::String(Aws::Region::US_EAST_1);
}
```
```
Aws::String EC2MetadataClient::GetCurrentRegion() const
        {
            if (!m_region.empty())
            {
                return m_region;
            }

            AWS_LOGSTREAM_TRACE(m_logtag.c_str(), "Getting current region for ec2 instance");

            Aws::StringStream ss;
            ss << m_endpoint << EC2_REGION_RESOURCE;
            std::shared_ptr<HttpRequest> regionRequest(CreateHttpRequest(ss.str(), HttpMethod::HTTP_GET,
                                                                         Aws::Utils::Stream::DefaultResponseStreamFactoryMethod));
            {
                std::lock_guard<std::recursive_mutex> locker(m_tokenMutex);
                if (m_tokenRequired)
                {
                    regionRequest->SetHeaderValue(EC2_IMDS_TOKEN_HEADER, m_token);
                }
            }
```
I have a test results, for the query below the time cost can optimized from 200s+ to < 10s. The reason is that we can increase the concurrency and we remove the time-cost operation during ClientConfiguration initialization  after this pr
```
MySQL [iceberg_10t]> SELECT sum(LO_EXTENDEDPRICE * LO_DISCOUNT) AS revenue FROM lineorder_flat WHERE LO_ORDERDATE = 19970101 AND LO_DISCOUNT BETWEEN 1 AND 3 AND LO_QUANTITY < 25;
```
you can also refer to:
https://github.com/aws/aws-sdk-cpp/issues/1511
https://github.com/aws/aws-sdk-cpp/issues/1440

### CODE REVIEW

File: be/src/env/env_s3.cpp [hunk 1]
```diff
@@ -194,7 +195,15 @@ S3ClientFactory::S3ClientPtr S3ClientFactory::new_client(const ClientConfigurati
 }
 
 static std::shared_ptr<Aws::S3::S3Client> new_s3client(const S3URI& uri) {
-    Aws::Client::ClientConfiguration config;
+    // Since aws sdk has changed the Aws::Client::ClientConfiguration default constructor to
+    // search for the region (where as before 1.8 it has been hard coded default of "us-east-1").
+    // Part of that change is looking through the ec2 metadata, which can take a long time.
+    // By having AWS_EC2_METADATA_DISABLED it avoids doing this search.
+    // For more details, please refer https://github.com/aws/aws-sdk-cpp/issues/1440
+    setenv("AWS_EC2_METADATA_DISABLED", "true", 1);
```
Comment (sduzh) [inline be/src/env/env_s3.cpp:203]: Is `setenv` thread-safe?

Comment (sduzh) [reply be/src/env/env_s3.cpp:203]: Can we skip searching region by constructing a `ClientConfiguration` instance by copying from a global variable?
```
static Aws::Client::ClientConfiguration s_config;
Aws::Client::ClientConfiguration config = s_config;
...
```

Comment (caneGuy) [reply be/src/env/env_s3.cpp:203]: > Can we skip searching region by creating a copy of a static variable of type Aws::Client::ClientConfiguration?
> 
> ```
> static Aws::Client::ClientConfiguration s_config;
> Aws::Client::ClientConfiguration config = s_config;
> ...
> ```

The config.endpointOverride field will be override if we use the same ClientConfiguration object for all endpoint.
I am thinking about cache static ClientConfiguration used endpoint as key.

**Comment by dirtysalt** [comment]
@caneGuy @sduzh  Maube you can do some initialization at the entry of main funciton `starrocks_main.cpp`, at 

```
#ifdef STARROCKS_WITH_AWS
    Aws::SDKOptions aws_sdk_options;
    Aws::InitAPI(aws_sdk_options);
#endif

    while (!starrocks::k_starrocks_exit) {
        sleep(10);
    }
    daemon->stop();
    daemon.reset();

#ifdef STARROCKS_WITH_AWS
    Aws::ShutdownAPI(aws_sdk_options);
#endif
```

File: be/src/env/env_s3.cpp (hunk 1 above)
Comment (sduzh) [reply be/src/env/env_s3.cpp:203]: > > Can we skip searching region by creating a copy of a static variable of type Aws::Client::ClientConfiguration?
> > ```
> > static Aws::Client::ClientConfiguration s_config;
> > Aws::Client::ClientConfiguration config = s_config;
> > ...
> > ```
> 
> The config.endpointOverride field will be override if we use the same ClientConfiguration object for all endpoint. I am thinking about cache static ClientConfiguration use endpoint as key.

We are not using the same object, each time you call `Aws::Client::ClientConfiguration config = s_config;` a new object is created and the `endpointOverride` will not be shared among each other.

**Comment by sduzh** [comment]
> 

I'm not familiar with AWS SDK., initialize like on start can skip the process of finding current region each time create a ClientConfiguration? 

> @caneGuy @sduzh Maube you can do some initialization at the entry of main funciton `starrocks_main.cpp`, at
> 
> ```
> #ifdef STARROCKS_WITH_AWS
>     Aws::SDKOptions aws_sdk_options;
>     Aws::InitAPI(aws_sdk_options);
> #endif
> 
>     while (!starrocks::k_starrocks_exit) {
>         sleep(10);
>     }
>     daemon->stop();
>     daemon.reset();
> 
> #ifdef STARROCKS_WITH_AWS
>     Aws::ShutdownAPI(aws_sdk_options);
> #endif
> ```

I'm not familiar with AWS SDK, do you mean by doing this we can avoid getting the current region each time construct a ClientConfiguration object?

**Comment by dirtysalt** [comment]
Yes. Two options:
1. set region in ENV explicitly.
2. or disable 

image

**Comment by dirtysalt** [comment]
We can get running region before staring be, set environment variable like `AWS_REGION`. But I'm not sure if there is equivalence in Aliyun.  


- https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
- https://gist.github.com/quiver/87f93bc7df6da7049d41

image

**Comment by sduzh** [comment]
> 

This won't fix the problem if we're running BE in a physical machine, not hosted by cloud vender.

**Comment by dirtysalt** [comment]
> > 
> 
> This won't fix the problem if we're running BE in a physical machine not hosted by cloud vender.

Then we can fallback to US_EAST_1.

**Comment by dirtysalt** [comment]
Although I'm not 100% sure, I guess the way of sdk to tell current region, is to raise a http request to  `http://169.254.169.254/latest/meta-data/`

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html

`EC2_REGION_RESOURCE` in code is  `static const char EC2_REGION_RESOURCE[] = "/latest/meta-data/placement/availability-zone";`


image

**Comment by sduzh** [comment]
>
> 
> > > 
> > 
> > 
> > This won't fix the problem if we're running BE in a physical machine not hosted by cloud vender.
> 
> Then we can fallback to US_EAST_1.

It's possible we are connecting to an S3 compatible service not provided by the AWS, Assigning US_EAST_1 as the default region may not work, but I'm not sure about this.

On the other hand, if BE is running on an EC2 machine, maybe there will be no need to detect the current region  in `start_be.sh` at all, because the ClientConfiguration can detect the region by itself and once the region has been detected it will be cached for later use, the above-mentioned exclusive lock will not be accessed anymore.

**Comment by dirtysalt** [comment]
> > 
> 
> This won't fix the problem if we're running BE in a physical machine not hosted by cloud vender.

Physical machine can not determine the best region to s3, so it has to fallback to default region.

I guess we can have a better idea:
1. Let user specify environment like `AWS_REGION`
2. In start_be.sh, if we don't have `AWS_REGION`, we can search by ourselves.

**Comment by dirtysalt** [comment]
> > > > 
> > > 
> > > 
> > > This won't fix the problem if we're running BE in a physical machine not hosted by cloud vender.
> > 
> > 
> > Then we can fallback to US_EAST_1.
> 
> It's possible we are connecting to an S3 compatible service not provided by the AWS, Assigning US_EAST_1 as the default region may not work, but I'm not sure about this.
> 
> On the other hand, if BE is running on an EC2 machine, maybe there will be no need to detect the current region in `start_be.sh` at all, because the ClientConfiguration can detect the region by itself and once the region has been detected it will be cached for later use, the above-mentioned exclusive lock will not be accessed anymore.

Yes, `ClientConfiguration` can detect it, and we just do it explicitly outside of Starrocks. Maybe much more reliable and controllable.

I'm okay with your solution if current code can avoid lock contention(cache or any technique)

**Comment by dirtysalt** [comment]
> > 
> 
> IMO, this would make things more complicated, simply cache a ClientConfiguration in heap memory and make a deep copy of it each time you want to create a S3Client would solve the problem.

Yes, I think it's okay if implement is simple enough.

**Comment by sduzh** [comment]
> 

And you can still set the environment `AWS_REGION` as you want, the environment variable will be used to initialize the global-shared cached ClientConfiguration.

File: be/src/env/env_s3.cpp (hunk 1 above)
Comment (caneGuy) [reply be/src/env/env_s3.cpp:203]: > > > Can we skip searching region by creating a copy of a static variable of type Aws::Client::ClientConfiguration?
> > > ```
> > > static Aws::Client::ClientConfiguration s_config;
> > > Aws::Client::ClientConfiguration config = s_config;
> > > ...
> > > ```
> > 
> > 
> > The config.endpointOverride field will be override if we use the same ClientConfiguration object for all endpoint. I am thinking about cache static ClientConfiguration use endpoint as key.
> 
> We are not using the same object, each time you call `Aws::Client::ClientConfiguration config = s_config;` a new object is created and the `endpointOverride` will not be shared among each other.

LGTM，i will change the implementation

**Comment by dirtysalt** [comment]
I don't expect implemention could be so simple!!!

**Review by sduzh — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Comment by dirtysalt** [comment]
Failed to start_be, crash at. I guess clientconfiguration must be initliaize after calling 
```
    Aws::SDKOptions aws_sdk_options;
    Aws::InitAPI(aws_sdk_options);
```

image

**Comment by sduzh** [comment]
https://github.com/StarRocks/starrocks/issues/3927

**Pull request merged**